File: python/game/Snake.py
import copy

from python.game.enums import Direction
from python.game.globals import SNAKE_MAX_HEALTH, SNAKE_STARTER_SIZE
from python.game.Point import Point


class Snake:
    def __init__(self, ID):
        self.health = SNAKE_MAX_HEALTH
        self.body = []
        self.name = ""
        self.ID = ID
        self.size = SNAKE_STARTER_SIZE
        self.alive = True

    def assign_start_position(self, start_position: Point):
        for _ in range(self.size):
            self.body.append(start_position)

    @property
    def head(self) -> Point:
        return self.body[0]

    def all_possible_moves(self):
        if not self.body:
            return []

        return [
            Point(self.head.x - 1, self.head.y),
            Point(self.head.x + 1, self.head.y),
            Point(self.head.x, self.head.y - 1),
            Point(self.head.x, self.head.y + 1),
        ]

    def all_positions_at(self, distance: int):
        return [
            Point(self.head.x - distance, self.head.y - distance),
            Point(self.head.x + distance, self.head.y + distance),
            Point(self.head.x - distance, self.head.y - 1 + distance),
            Point(self.head.x + distance, self.head.y + 1 - distance),
        ]

    def move(self, direction: Direction):
        new_head = copy.deepcopy(self.head)
        if direction == Direction.DOWN:
            new_head.y += 1
        elif direction == Direction.LEFT:
            new_head.x -= 1
        elif direction == Direction.RIGHT:
            new_head.x += 1
        elif direction == Direction.UP:
            new_head.y -= 1
        else:
            dx = self.body[0].x - self.body[1].x
            dy = self.body[0].y - self.body[1].y
            if dx == 0 and dy == 0:
                dy = 1
            new_head.x += dx
            new_head.y += dy

        self.body.insert(0, new_head)
        self.body.pop()

    def feed(self):
        self.health = SNAKE_MAX_HEALTH
        self.body.append(self.body[-1])

    def colliding_with(self, snake):
        for body_part in snake.body:
            if body_part == snake.head:
                continue

            if body_part == self.head:
                return True
        return False

File: express/game/snake.js
const _ = require('lodash');

const {
  DIRECTION
} = require("./enums.js");

const {
  SNAKE_MAX_HEALTH,
  SNAKE_STARTER_SIZE
} = require("./globals.js");

const Point = require('./point.js');

function Snake(ID) {
  this.health = SNAKE_MAX_HEALTH;
  this.body = [];
  this.name = "";
  this.ID = ID;
  this.size = SNAKE_STARTER_SIZE;
  this.alive = true;

  function assign_start_position(start_position) {
    _.times(this.size, () => this.body.append(start_position));
  }
  
  Object.defineProperties(this, {
    head: {
      get: () => this.body[0],
      enumerable: true
    }
  });

  function all_possible_moves() {
    if(!this.body)
      return [];

    return [
      new Point(this.head.x - 1, this.head.y),
      new Point(this.head.x + 1, this.head.y),
      new Point(this.head.x, this.head.y - 1),
      new Point(this.head.x, this.head.y + 1),
    ];
  }

  function all_positions_at(distance) {
    return [
      new Point(this.head.x - distance, this.head.y - distance),
      new Point(this.head.x + distance, this.head.y + distance),
      new Point(this.head.x - distance, this.head.y - 1 + distance),
      new Point(this.head.x + distance, this.head.y + 1 - distance),
    ];
  }

  function move(direction) {
    var new_head = new Point(this.head.x, this.head.y);
    if(direction == DIRECTION.DOWN)
        new_head.y += 1;
    else if(direction == DIRECTION.LEFT)
        new_head.x -= 1;
    else if(direction == DIRECTION.RIGHT)
        new_head.x += 1;
    else if(direction == DIRECTION.UP)
        new_head.y -= 1;
    else {
      var dx = this.body[0].x - this.body[1].x;
      var dy = this.body[0].y - this.body[1].y;
      if(dx == 0 && dy == 0)
          dy = 1;
      new_head.x += dx;
      new_head.y += dy;
    }

    this.body.splice(0, 0, new_head);
    this.body.pop();
  }

  function feed() {
    this.health = SNAKE_MAX_HEALTH
    this.body.push(this.body[this.body.length - 1]);
  }

  function colliding_with(snake) {
    snake.body.forEach(body_part => {
      if(body_part.equals(snake.head))
        return;

      if(body_part.equals(this.head))
        return true;
    });
    return false;
  }
  
  Object.assign(this, {
    assign_start_position,
    all_possible_moves,
    all_positions_at,
    move,
    feed,
    colliding_with
  });
}

module.exports = Snake;